Add about section

AndycptSher opened the pull request:
#12
implemented about section

AndycptSher committed:
Typo
compair -> compare

AndycptSher committed:
Minor Refactor
update to not check for component castable in metadata sections

AndycptSher committed:
Add about section
- created new file 'spell_attributes_list' of all attributes and their associated opcodes
- refactored 'parse_spell' to be able to identify current section
- added 'about' section
- refactored to disallow component parsing while in 'about' section
- added 'parse_equation'
- added tests for 'about' section
- added 'set_meta_data' method to 'Spell'

AndycptSher committed:
Removed finished TODO

CocytusDEDI commented on docs/spell_attributes_list.txt:
```diff
@@ -0,0 +1,4 @@
+TODO: format
+
```
```suggestion
```

CocytusDEDI commented on src/lib.rs:
```diff
@@ -636,10 +636,17 @@ impl Spell {
         while let Some(&bits) = instructions_iter.next() {
+            if let Some(502) = section { // ignore all checks in metadata section
+                continue;
+            }
```
Section can never be rewritten to after section becomes 502

CocytusDEDI commented on src/lib.rs:
```diff
@@ -719,10 +729,38 @@ impl Spell {
     }
 
+    fn set_meta_data(&mut self, equations: Vec<u64>) {
+        let mut codes = equations.into_iter();
```
Equations should be renamed to attributes

```suggestion
    fn set_meta_data(&mut self, attributes: Vec<u64>) {
        let mut codes = attributes.into_iter();
```

CocytusDEDI commented on src/lib.rs:
```diff
@@ -719,10 +729,38 @@ impl Spell {
+                1 => { // set destroy_when_done
+                    codes.next();
+                    panic!("Implementation not specified for destroy_when_done")
+                },
```
destroy_when_done is planned yet
```suggestion
```

CocytusDEDI commented on src/lib.rs:
```diff
@@ -719,10 +729,38 @@ impl Spell {
+                    .map(|x| f64::from_bits(x) as f32)
+                    .collect::<Vec<f32>>()[..] {
+                        [c, d, e] => self.color = Color{r: c, g: d, b: e, a: SPELL_TRANSPARENCY},
+                        _ => panic!("Invalid data")
```
```suggestion
                0 => { // Set colour
                    match match vec![codes.next(), codes.next(), codes.next()].into_iter().collect::<Option<Vec<u64>>>(){ // Transpose vec of option into option of vec
                        Some(color_vector) => color_vector,
                        None => panic!("Invalid data: There should be three color values")
                    }.into_iter()
                    .map(|x| f64::from_bits(x) as f32)
                    .collect::<Vec<f32>>()[..] {
                        [red, green, blue] => self.color = Color{r: red, g: green, b: blue, a: SPELL_TRANSPARENCY},
                        _ => panic!("Failed to parse colors")
```

CocytusDEDI commented on src/lib.rs:
```diff
@@ -719,10 +729,38 @@ impl Spell {
+                _ => {}
+            }
+        }
+
```
```suggestion
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -1,11 +1,13 @@
 use lazy_static::lazy_static;
 use std::collections::HashMap;
+use core::num::ParseFloatError;
```
Don't use ParseFloatError, as the rest of the program just uses `&'static str`

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
     }
 }
 
+fn parse_equation(equation: &str) -> Result<Vec<u64>, &'static str>{
```
Better name needed
```suggestion
fn parse_about_line(equation: &str) -> Result<Vec<u64>, &'static str>{
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -80,6 +83,8 @@ pub fn parse_spell(spell_code: &str) -> Result<Vec<u64>, &'static str> {
             } else if trimmed_line == "" {
                 continue
+            } else if trimmed_line.contains('=') && Some(502) == in_section { // Indicates an assignment of metadata
+                instructions.extend(parse_equation(trimmed_line)?)
```
```suggestion
                instructions.extend(parse_about_line(trimmed_line)?)
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
 
+fn parse_equation(equation: &str) -> Result<Vec<u64>, &'static str>{
+    let (name, value) = match equation.split_once('='){
+        Some(n) => n,
```
Name variables better, shouldn't be just `n`

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
+        None => return Err("equation not valid")
+    };
+    
+    match (name.trim(), value.trim()) { // because no map for tuples in rust
```
```suggestion
    match (name.trim(), value.trim()) {
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
+            }.split(',')
+            .map(str::trim)
+            .map(str::parse::<f32>)
+            .collect::<Result<Vec<f32>, ParseFloatError>>() {
```
Don't use ParseFloatError

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
+            .map(str::trim)
+            .map(str::parse::<f32>)
+            .collect::<Result<Vec<f32>, ParseFloatError>>() {
+                Ok(n) => n,
```
Name variables better than just `n`

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
+    match (name.trim(), value.trim()) { // because no map for tuples in rust
+        ("colour", values) | ("color", values) => {
+            // returns used to sidestep borrowing rules
+            let numbers = match match match match values.strip_prefix('[')
```
Having four match statements is insane behaviour, and makes it very unreadable. This needs to be refactored.

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -659,6 +664,53 @@ fn parse_parameter(parameter_string: &str, parameter_type: u64) -> Result<Parame
+            "true" => Ok(vec![1, 1]),
+            "false" => Ok(vec![1, 0]),
+            _ => Err("Invalid parameter")
+        },
```
destroy_when_done isn't planned yet
```suggestion
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -694,6 +746,33 @@ mod tests {
+    }
+
+    #[test]
+    fn parse_limit_colour_attribution() {
```
```suggestion
    fn parse_limit_colour() {
```

CocytusDEDI commented on src/spelltranslator.rs:
```diff
@@ -694,6 +746,33 @@ mod tests {
+        assert_eq!(parse_equation("     color      =        [   0.212,    1,0.3]"), Ok(vec![0, f64::to_bits((0.212 as f32) as f64), f64::to_bits((1 as f32) as f64), f64::to_bits((0.3 as f32) as f64)]));
+    }
+
+    #[test]
+    fn parse_bool_attrtbution() {
+        assert_eq!(parse_equation("destroy_when_done = true"), Ok(vec![1, 1]))
+    }
```
```suggestion
```

CocytusDEDI reviewed (changes requested):
Overall major refactors need to be done for readability and one bug fix. After it has been refactored testing will be done.

AndycptSher committed:
Refactored after review
- removed destroy_when_done
- clarified variable names
- renamed functions to be better named

AndycptSher committed:
Patch
refactored function names forgotten in previous commit

AndycptSher committed:
Refactor after review

AndycptSher commented:
## Note: possible bug
current implementation permits if statements in about section introducing possible parsing errors as additional opcodes have not been accounted for

AndycptSher committed:
Merge branch 'main' into main

CocytusDEDI commented:
Can confirm that's a bug, I'll be fixing it along with a couple other things.
